backend/src/android/androidIntegrity.ts

```diff
@@ -0,0 +1,169 @@
+ * The native implementation contacates the chain of certificates into a single base64 encoded string divided by commas.
+ * {@link https://developer.android.com/training/articles/security-key-attestation}
+ * 1st and 2nd steps are not implemented here as they are done on the client side.
+ * Functions used to verify the attestation have their relative steps commented in the TSdoc.
+ * Even though some steps might be joined in a single function, they are separated in the comments for clarity at the cost of some repetition.
+ * The 5th step is not implemented as it is only implemented in newer chains.
+ * The 7th step which compares the attestation extension data with a fixed set of expected values is not implemented in this example as it heavily depends on the specific use case.
+ * @param x509Array - The chain of X.509 certificates in base64 format divided by commas.
+ * @throws {Error} - If the attestation is invalid.
+ */
+export const verifyAttestation = async (x509Array: string) => {
+  // We split the chain of certificates and convert them to PEM format to be parsed by the X509Certificate class
+  const decodedString = Buffer.from(x509Array, 'base64').toString('utf-8');
+  const certificates = decodedString.split(',');
+  const x509Chain = certificates.map((cert) => {
+    return new X509Certificate(base64ToPem(cert));
+  });
+  validateIssuance(x509Chain);
+  await validateRevokation(x509Chain);
+  validateKeyAttestationExtension(x509Chain);
+};
+
+/**
+ * 3.
+ * Obtain a reference to the X.509 certificate chain parsing and validation library that is most appropriate for your toolset.
+ * Verify that the root public certificate is trustworthy and that each certificate signs the next certificate in the chain.
+ * @param x509Chain - The chain of {@link X509Certificate} certificates.
+ * @throws {Error} - If the chain is invalid.
+ */
+const validateIssuance = (x509Chain: X509Certificate[]) => {
+  if (x509Chain.length === 0) throw new Error('No certificates provided');
+
+  // Check dates
+  const now = new Date();
+  const datesValid = x509Chain.every(
+    (c) => new Date(c.validFrom) < now && now < new Date(c.validTo)
+  );
+  if (!datesValid) throw new Error('Certificates expired');
+
+  // Check that each certificate, except for the last, is issued by the subsequent one.
+  if (x509Chain.length >= 2) {
+    for (let i = 0; i < x509Chain.length - 1; i++) {
+      const subject = x509Chain[i];
+      const issuer = x509Chain[i + 1];
+
+      if (
+        !subject ||
+        !issuer ||
+        subject.checkIssued(issuer) === false ||
+        subject.verify(issuer.publicKey) === false
+      ) {
+        throw new Error('Certificate chain is invalid');
+      }
+    }
+  }
+
+  // Ensure that the last certificate in the chain is the expected Google Hardware Attestation Root CA.
+  const pkFile = fs.readFileSync(
+    require('path').resolve(__dirname, 'googleHardwareAttestationRoot.key')
```

Comment:
Should we put `'googleHardwareAttestationRoot.key'` in a `.env` variable? What do you think?

Reply:
I'd rather keep it in a file as it's something not dependant from the environment, it's fixed until it gets revoked and at that point swapping the file works without rebuilding with a new env. Does it make sense?